File: frontend/src/api/results.ts
import { api } from "./client";

export interface MarksheetPaper {
  code: string | null;
  name: string | null;
  type: string | null;
  max: string | null;
  theory: string | null;
  practical: string | null;
  internal: string | null;
  total: string | null;
  status: string | null;
  grade: string | null;
  credit: string | null;
  point: string | null;
}

export interface Marksheet {
  rollno: string | null;
  enroll_no: string | null;
  name: string | null;
  father_name: string | null;
  course_id: string | null;
  course_name: string | null;
  semester: string | null;
  college_name: string | null;
  category: string | null;
  exam_month: string | null;
  marksheet_no: string | null;
  papers: MarksheetPaper[];
  sgpa: string | null;
  result: string | null;
  grand_cgpa: string | null;
  grand_percent: string | null;
  grand_division: string | null;
  grand_result: string | null;
  total: string | null;
  max: string | null;
}

export async function getMarksheets(rollno: string): Promise<Marksheet[]> {
  const { data } = await api.get<Marksheet[]>(`/results/${encodeURIComponent(rollno)}`);
  return data;
}

export async function myResult(): Promise<Marksheet[]> {
  const { data } = await api.get<Marksheet[]>("/student/result");
  return data;
}

export async function openMarksheetPdf(rollno: string): Promise<void> {
  const resp = await api.get(`/results/${encodeURIComponent(rollno)}/pdf`, { responseType: "blob" });
  const url = URL.createObjectURL(resp.data as Blob);
  window.open(url, "_blank");
}

File: backend/app/schemas/results.py
"""Schemas for results / marksheet."""
from pydantic import BaseModel


class MarksheetPaper(BaseModel):
    code: str | None = None
    name: str | None = None
    type: str | None = None
    max: str | None = None
    theory: str | None = None
    practical: str | None = None
    internal: str | None = None
    total: str | None = None
    status: str | None = None
    grade: str | None = None
    credit: str | None = None
    point: str | None = None


class Marksheet(BaseModel):
    rollno: str | None = None
    enroll_no: str | None = None
    name: str | None = None
    father_name: str | None = None
    course_id: str | None = None
    course_name: str | None = None
    semester: str | None = None
    college_name: str | None = None
    category: str | None = None
    exam_month: str | None = None
    marksheet_no: str | None = None
    papers: list[MarksheetPaper] = []
    sgpa: str | None = None
    result: str | None = None
    grand_cgpa: str | None = None
    grand_percent: str | None = None
    grand_division: str | None = None
    grand_result: str | None = None
    total: str | None = None
    max: str | None = None
